crates/block-producer/src/block_builder/prover/block_witness.rs

```diff
@@ -31,44 +33,13 @@ pub struct BlockWitness {
 
 impl BlockWitness {
     pub fn new(
-        block_inputs: BlockInputs,
+        mut block_inputs: BlockInputs,
         batches: &[TransactionBatch],
-    ) -> Result<Self, BuildBlockError> {
-        Self::validate_inputs(&block_inputs, batches)?;
-
-        let updated_accounts = {
-            let mut account_initial_states: BTreeMap<AccountId, Digest> =
-                batches.iter().flat_map(TransactionBatch::account_initial_states).collect();
-
-            let mut account_merkle_proofs: BTreeMap<AccountId, MerklePath> = block_inputs
-                .accounts
-                .into_iter()
-                .map(|(account_id, witness)| (account_id, witness.proof))
-                .collect();
-
-            batches
-                .iter()
-                .flat_map(TransactionBatch::updated_accounts)
-                .map(|update| {
-                    let initial_state_hash = account_initial_states
-                        .remove(&update.account_id())
-                        .expect("already validated that key exists");
-                    let proof = account_merkle_proofs
-                        .remove(&update.account_id())
-                        .expect("already validated that key exists");
-
-                    (
-                        update.account_id(),
-                        AccountUpdateWitness {
-                            initial_state_hash,
-                            final_state_hash: update.new_state_hash(),
-                            proof,
-                            transactions: update.transactions().to_vec(),
-                        },
-                    )
-                })
-                .collect()
-        };
+    ) -> Result<(Self, Vec<BlockAccountUpdate>), BuildBlockError> {
+        if batches.len() > MAX_BATCHES_PER_BLOCK {
+            return Err(BuildBlockError::TooManyBatchesInBlock(batches.len()));
+        }
+        Self::validate_nullifiers(&block_inputs, batches)?;
```

Comment:
The validate function became a bit of a liability - it can no longer be performed at the start but instead needs to be done in parts since account state can only be checked after we've merged the batch updates.